python/indexing/mlir_indexing/dialects/indexing.py

```diff
@@ -0,0 +1,63 @@
+#  Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+#  See https://llvm.org/LICENSE.txt for license information.
+#  SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+import atexit
+import contextlib
+import ctypes
+import sys
+from typing import Optional, Sequence
+
+
+# this is specifically for _ZN4mlir6python23DefaultingPyMlirContext7resolveEv
+@contextlib.contextmanager
+def dl_open_guard():
+  old_flags = sys.getdlopenflags()
+  sys.setdlopenflags(old_flags | ctypes.RTLD_GLOBAL)
+  yield
+  sys.setdlopenflags(old_flags)
+
+
+with dl_open_guard():
+  # noinspection PyUnresolvedReferences
+  from .._mlir_libs import _mlir
+  from ..ir import (register_attribute_builder, DenseI64ArrayAttr, Context,
+                    Location, Module, InsertionPoint)
```

Comment:
Adding the attribute `__attribute__ ((visibility("default")))` does what it says - makes the symbol visibility _default_, not _public_ (with the attribute they are "hidden"). On Linux the typeinfo (re [RTTI](https://en.cppreference.com/w/cpp/types) that pybind employs) appear in the symbol table in the ELF but they are "weak" symbols

```console
$ readelf -s -W -C _mlir.cpython-310-x86_64-linux-gnu.so | grep typeinfo | grep PyOperation
687: 000000000036c4e8    56 OBJECT  WEAK   DEFAULT   23 typeinfo for mlir::python::PyOperation
635: 0000000000260bf0    28 OBJECT  WEAK   DEFAULT   16 typeinfo name for mlir::python::PyOperation
```
or "vague" symbols:

```console
$ nm -gDC _mlir.cpython-310-x86_64-linux-gnu.so | grep typeinfo | grep PyOperation
000000000036c4e8 V typeinfo for mlir::python::PyOperation
0000000000260bf0 V typeinfo name for mlir::python::PyOperation
```

That means that means they are not available for symbol resolution/relocation of subsequently loaded shared objects ([man dlopen](https://linux.die.net/man/3/dlopen)), i.e., a C extension that's loaded after the upstream `_mlir.cpython` extension is loaded. This corresponds to a default loader flag `RTLD_LOCAL`; `RTLD_GLOBAL` makes them available.

Yes I have spent an inordinate amount of time fighting linkers (even [just this week](https://reviews.llvm.org/D148058)).